numpy/core/src/common/simd/avx512/arithmetic.h

```diff
@@ -112,5 +115,27 @@ NPY_FINLINE __m512i npyv_mul_u8(__m512i a, __m512i b)
 // TODO: emulate integer division
 #define npyv_div_f32 _mm512_div_ps
 #define npyv_div_f64 _mm512_div_pd
+NPY_FINLINE float npyv_sum_f32(npyv_f32 a)
+{
+    __m512 h64   = _mm512_shuffle_f32x4(a, a, _MM_SHUFFLE(3, 2, 3, 2));
+    __m512 sum32 = _mm512_add_ps(a, h64);
+    __m512 h32   = _mm512_shuffle_f32x4(sum32, sum32, _MM_SHUFFLE(1, 0, 3, 2));
+    __m512 sum16 = _mm512_add_ps(sum32, h32);
+    __m512 h16   = _mm512_permute_ps(sum16, _MM_SHUFFLE(1, 0, 3, 2));
+    __m512 sum8  = _mm512_add_ps(sum16, h16);
+    __m512 h4    = _mm512_permute_ps(sum8, _MM_SHUFFLE(2, 3, 0, 1));
+    __m512 sum4  = _mm512_add_ps(sum8, h4);
+    return _mm_cvtss_f32(_mm512_castps512_ps128(sum4));
+}
+NPY_FINLINE double npyv_sum_f64(npyv_f64 a)
+{
+    __m512d h64   = _mm512_shuffle_f64x2(a, a, _MM_SHUFFLE(3, 2, 3, 2));
+    __m512d sum32 = _mm512_add_pd(a, h64);
+    __m512d h32   = _mm512_permutex_pd(sum32, _MM_SHUFFLE(1, 0, 3, 2));
+    __m512d sum16 = _mm512_add_pd(sum32, h32);
+    __m512d h16   = _mm512_permute_pd(sum16, _MM_SHUFFLE(2, 3, 0, 1));
+    __m512d sum8  = _mm512_add_pd(sum16, h16);
+    return _mm_cvtsd_f64(_mm512_castpd512_pd128(sum8));
+}
```

Comment:
If you have a link indicating that this is the most efficient way to do this sum, it would be great to include it in a comment.

Reply:
The blame on me. currently, I would say there are two ways to implement reduce sum for AVX512 go with **hadd** and just **add** for example:

1-  split(256)     /add  /split(128)     /add  /hadd /hadd /extract
2-  shuff(cross) /add  /shuff(cross) /add  /shuff /add   /shuff   /add  /extract

The first one is been widely used by many projects while the second one is used by Intel Compiler and by this patch,
the reason why the second preferred by intel compiler(My guess) is that the latency of hadd increased by (2-3)
starting from Skylake-X which makes two extra shuffles(non-cross) cheaper. check [Godbolt](https://godbolt.org/z/s3G9Er) for more clarification.

Reply:
Compared with the following code:
```
float npyv_sum_f32(__m512 a)
{
    __m256 t0 = _mm256_add_ps(_mm512_castps512_ps256(a), 
    _mm256_castpd_ps(_mm512_extractf64x4_pd(_mm512_castps_pd(a),1)));
    __m128 t1 = _mm_add_ps(_mm256_castps256_ps128(t0), _mm256_extractf128_ps(t0,1));
    __m128 t2 = _mm_movehdup_ps(t1);
    __m128 t3 = _mm_add_ps(t1, t2);
    __m128 t4 = _mm_movehl_ps(t3, t3);
    __m128 t5 = _mm_add_ss(t3, t4);
    return _mm_cvtss_f32(t5);
}
```
The current shuffle method is faster.

Reply:
>  the second one is used by Intel Compiler and by this patch,

Please add this in a comment in the file!

Reply:
I'm confused by your godbolt link. Why don't we just use `_mm512_reduce_add_ps`?

Reply:
As you can see in godbolt, This implementation almost the same as `_mm512_reduce_add_ps/pd` but works for all compilers. `_mm512_reduce_add_ps` only works on intel compilers.

Reply:
Does it make sense to `#ifdef` intel compilers and use the intrinsic in them?

Reply:
GCC 7.1 onwards has this intrinsic, as does clang 4.

Reply:
We still need to support older compilers.

Reply:
True, but we don't need to optimize for them if it's easier not to. In order of preference, I think we should:

* `#ifdef` to use the intrinsic if available, the code in this patch as fallback
* `#ifdef` to use the intrinsic if available, unoptimized code if not
* Keep this patch as is

Reply:
@eric-wieser,
Both `_mm512_reduce_add_ps/pd` are not widely supported and using them within C directive `#if
will require an independent test against the compilers including Intel compilers,
also on clang (<7) `_mm512_reduce_add_ps/pd` are slower due to using `vhadd`.

I think it's fine to keep this implementation as-is since these intrinsics don't have a native support
by the machine and the compilers emulated them anyway.

Reply:
> will require an independent test against the compilers including Intel compilers

Can we test them as part of `config_test.c` to emit a `has_mm512_reduce_add_ps` macro?

> also on clang (<7) `_mm512_reduce_add_ps/pd` are slower due to using `vhadd`.

I don't think we need to care about performance due to outdated dependencies.

Reply:
> Can we test them as part of config_test.c to emit a has_mm512_reduce_add_ps macro?

yes as a temporary solution since that will break the portability plan, it's better to extend `CCompilerOpt` to test intrinsics too.

> I don't think we need to care about performance caused by outdated dependencies.

I agree.

Reply:
Let's Keep this patch as is with the detailed comments attached.